### PR TITLE
cherry picked partitioner

### PR DESCRIPTION
**Pull request by alan-stokes**
changes from roberts conversion for allowing a application vertex to partiton itself. includes support for multiple types of vertex with different projection mappings

tied to
PAC: https://github.com/SpiNNakerManchester/PACMAN/pull/260
FEC: 
SPYNNAKER: https://github.com/SpiNNakerManchester/sPyNNaker/pull/708
SPYNNAKER8: https://github.com/SpiNNakerManchester/sPyNNaker8/pull/273

### CODE REVIEW

**Review by dkfellows — APPROVED** [review]
There might be a better way of doing the splitter selection, but the way used will work and this PR should not be delayed by waiting for the best solution when a good enough solution is already there.

**Comment by Christian-B** [comment]
> There might be a better way of doing the splitter selection, but the way used will work and this PR should not be delayed by waiting for the best solution when a good enough solution is already there.

There are lots of ideas of better way to do splitter selection but they are out of scope of this PR.

File: spinn_front_end_common/utility_models/data_speed_up_packet_gatherer.py [hunk 1]
```diff
@@ -13,14 +13,15 @@
...
+from pacman.model.graphs.application.abstract import (
+    AbstractOneAppOneMachineVertex)
 from .data_speed_up_packet_gatherer_machine_vertex import (
     DataSpeedUpPacketGatherMachineVertex)
 
 
 class DataSpeedUpPacketGather(AbstractOneAppOneMachineVertex):
     """ The gatherer for the data speed up protocols. Gatherers are only ever\
-        deployed on chips with an ethernet connection.
+            deployed on chips with an ethernet connection.
```
Comment (andrewgait) [inline spinn_front_end_common/utility_models/data_speed_up_packet_gatherer.py:24]: There appears to be an extra tab/spaces on this line.

**Review by andrewgait — APPROVED** [review]
These changes look fine to me.

File: spinn_front_end_common/interface/splitter_selectors/splitter_selector.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    splitter objects where required
+    """
+
+    NOT_KNOWN_APP_VERTEX_ERROR_MESSAGE = (
+        "The SplitterSelector has not seen the {} vertex before. Therefore "
+        "there is no known splitter to allocate to this app vertex and so "
+        "will use the SplitterSliceLegacy Splitter.")
+
+    def __call__(self, app_graph):
+        """ basic selector which puts the legacy splitter object on
```
Comment (dkfellows) [inline spinn_front_end_common/interface/splitter_selectors/splitter_selector.py:39]: backslash needed at end of this line

File: spinn_front_end_common/interface/splitter_selectors/splitter_selector.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+        :param ApplicationGraph app_graph: app graph
+        :rtype: None
+        """
+        for app_vertex in app_graph:
+            if app_vertex.splitter is None:
+                self.vertex_selector(app_vertex)
+
+    def vertex_selector(self, app_vertex):
+        """ main point for selecting a splitter object for a given app vertex.
```
Comment (dkfellows) [inline spinn_front_end_common/interface/splitter_selectors/splitter_selector.py:50]: Backslash needed at end of every non-terminal line in this paragraph _or_ blank line before next line. I think a blank line is probably better in this case.

File: spinn_front_end_common/abstract_models/impl/tdma_aware_application_vertex.py [hunk 4]
```diff
@@ -44,19 +44,22 @@ class TDMAAwareApplicationVertex(ApplicationVertex):
...
-    def __init__(self, label, constraints, max_atoms_per_core):
+    def __init__(self, label, constraints, max_atoms_per_core, splitter=None):
         """
         :param str label: The optional name of the vertex.
         :param iterable(~pacman.model.constraints.AbstractConstraint) \
                 constraints:
             The optional initial constraints of the vertex.
         :param int max_atoms_per_core: The max number of atoms that can be
             placed on a core, used in partitioning.
+        :type splitter: None or AbstractSplitterCommon
```
Comment (dkfellows) [inline spinn_front_end_common/abstract_models/impl/tdma_aware_application_vertex.py:55]: Need to check that the correct type is resolved. Stupid Sphinx…

File: spinn_front_end_common/interface/splitter_selectors/splitter_selector.py (hunk 2 above)
Comment (alan-stokes) [reply spinn_front_end_common/interface/splitter_selectors/splitter_selector.py:39]: fixed

File: spinn_front_end_common/interface/splitter_selectors/splitter_selector.py (hunk 3 above)
Comment (alan-stokes) [reply spinn_front_end_common/interface/splitter_selectors/splitter_selector.py:50]: fixed

**Pull request merged**